### PR TITLE
Add more robust reference parameters

### PR DESCRIPTION
**Pull request by btorresgil**
Includes link, find, findall, and delete_all for references

Works with new predefined object support

Notify: @lampwins

### CODE REVIEW

**Comment by lampwins** [comment]
I like where this is headed. I have not had a chance to do much actual testing with it. I do have few thoughts though. 

You are doing a recursive find for the objects which I think is going to be problematic when thinking about device group hierarchies. Furthermore, I don't think the library even considers the possibility of device group hierarchies currently. Which my 5 minutes of testing, it looks like it would not be a trivial thing to implement because the hierarchy is not constructed in the normal config tree. Object precedence with dev group hierarchy and even panorama vs local FW objects is going to effect how you search for objects. My initial thought is that you will have to recurse up the tree, not down it. Just thinking out loud.

Also, it looks like you are effectively overriding the strings/lists with the actual objects. Won't that effect the write functionality of the library or are you planning to handle that by just looking at the object name when building the xml tree?

File: pandevice/base.py [hunk 1]
```diff
@@ -1900,6 +1900,174 @@ def element_str(self):
...
+        if self.nearest_pandevice().parent:
+            ref_obj = self.panorama().find(name, var_path.references, recursive=True)
+            if ref_obj:
+                return ref_obj
+        # Couldn't find the referenced object anywhere in the config tree
+        # so check in the predefined objects
+        if include_predefined:
+            for reftype in var_path.references:
+                if reftype in self.nearest_pandevice().predefined.OBJECT_TYPES:
+                    self.nearest_pandevice().predefined.object(name, reftype)
```
Comment (shinmog) [inline pandevice/base.py]: Is this supposed to be `return self.nearest_pandevice().predefined.object(name, reftype)`?

File: pandevice/base.py [hunk 2]
```diff
@@ -2047,15 +2215,17 @@ class ParamPath(object):
...
     """
     def __init__(self, param, path=None, vartype=None,
-                 condition=None, values=None):
+                 condition=None, values=None, references=None):
         self.param = param
         self.path = path
         self.vartype = vartype
         self.condition = condition or {}
-        self.values = values or []
+        self.values = values or ()
```
Comment (shinmog) [inline pandevice/base.py]: Why is this an empty tuple now instead of an empty list?

File: pandevice/base.py [hunk 3]
```diff
@@ -2047,15 +2215,17 @@ class ParamPath(object):
...
+        self.references = references or ()
```
Comment (shinmog) [inline pandevice/base.py]: Multiple references..  This is for stuff like `pandevice.policies.SecurityRule`, where a source could be an freeform IP address, `pandevice.objects.AddressObject`, or `pandevice.objects.AddressGroup`..?

File: pandevice/base.py (hunk 2 above)
Comment (btorresgil) [inline pandevice/base.py]: It just didn't seem like the `values` for a param should be mutable so I made it a tuple.  But I haven't gone around to change the lists to tuples which would also need to be done.  What do you think, is there a reason to use a mutable list?

File: pandevice/base.py (hunk 3 above)
Comment (btorresgil) [inline pandevice/base.py]: That's correct.  A param could point to an AddressObject and an AddressGroup

**Comment by btorresgil** [comment]
@lampwins Yes, you're correct that this doesn't handle DG hierarchy accurately.  That is a whole other conversation because, as you say, it's not modified in the config API.  There is another concern that it could link an object with the same name in another vsys.  Recursing up instead of down could help, I'll have to think about that.  I'm not sure we can do anything about DG hierarchy until that is fully supported in the rest of the library.

Thanks for the input!  Will think about this and make some changes.

File: pandevice/base.py (hunk 2 above)
Comment (shinmog) [inline pandevice/base.py]: Nope, no reason for it to be mutable.  So just update the params that have `values` to be tuples and change the docstring from saying `values` is a `list` to it being a `tuple`.

**Comment by btorresgil** [comment]
Closing this for now. If we decide to improve the references, we'll use this as history, but will likely go another route.

**Pull request closed without merging**